### PR TITLE
Optimize Artery RemoteInstrument

### PR DESCRIPTION
**Pull request by bantonsson**
This is a cleanup and optimization of the RemoteInstrument code in Artery.

* Change RemoteInstrument API to work with ByteBuffers
* Remove the MetadataMap since serialization is done in buffer

Here are some unscientific benchmark numbers from my laptop.

```
----------------------------------------------------------------------------------------------------------------------
| Benchmark                                         |     (configType) |  Mode | Cnt |    Score |     Error |  Units |
|---------------------------------------------------|------------------|-------|-----|----------|-----------|--------|
| Old                                               |                  |       |     |          |           |        |
| CodecBenchmark.both                               |         Standard | thrpt |   5 |  750.592 | ± 103.790 | ops/ms |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space       |         Standard | thrpt |   5 |  475.762 | ± 608.715 | MB/sec |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space.norm  |         Standard | thrpt |   5 |  772.095 | ±  16.517 |   B/op |
| CodecBenchmark.both                               | RemoteInstrument | thrpt |   5 |  573.369 | ± 190.089 | ops/ms |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space       | RemoteInstrument | thrpt |   5 |  662.329 | ± 714.277 | MB/sec |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space.norm  | RemoteInstrument | thrpt |   5 | 1420.378 | ±  12.754 |   B/op |
| New                                               |                  |       |     |          |           |        |
| CodecBenchmark.both                               |         Standard | thrpt |   5 |  854.291 | ± 166.186 | ops/ms |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space       |         Standard | thrpt |   5 |  533.664 | ± 691.398 | MB/sec |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space.norm  |         Standard | thrpt |   5 |  758.389 | ±  20.873 |   B/op |
| CodecBenchmark.both                               | RemoteInstrument | thrpt |   5 |  791.343 | ± 170.640 | ops/ms |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space       | RemoteInstrument | thrpt |   5 |  516.096 | ± 652.720 | MB/sec |
| CodecBenchmark.both:·gc.churn.PS_Eden_Space.norm  | RemoteInstrument | thrpt |   5 |  778.509 | ±   8.818 |   B/op |
----------------------------------------------------------------------------------------------------------------------
```

The new RemoteInstrument is only slightly slower and consumes roughly the same number of bytes per op as the standard Artery code.

BTW, the materialization of the Streams still happen inside the measured part, even if I move everything else out. That happens several times during a single run. Would be nice to get that out of the way as well, but it's for another time.

### CODE REVIEW

**Comment by ktoso** [comment]
LGTM, I had reviewed this before.

**Review by ktoso — APPROVED** [review]

**Comment by bantonsson** [comment]
Any more opinions on this one?

**Comment by patriknw** [comment]
I'll review it.

File: akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala [hunk 1]
```diff
@@ -16,177 +20,272 @@ import akka.util.{ ByteString, OptionVal }
...
-    -1
+  def remoteWriteMetadata(recipient: ActorRef, message: Object, sender: ActorRef, buffer: ByteBuffer): Unit
 
-  override def remoteMessageSent(recipient: ActorRef, message: Object, sender: ActorRef): ByteString =
-    null
+  /**
+   * Called right before putting the message onto the wire.
+   * Parameters MAY be `null` (except `message` and `buffer`)!
+   */
+  def remoteMessageSent(recipient: ActorRef, message: Object, sender: ActorRef, size: Int, time: Long): Unit
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala]: add some doc about what `size` and `time` are and cross reference to `serializationTimingEnabled`

File: akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala [hunk 2]
```diff
@@ -16,177 +20,272 @@ import akka.util.{ ByteString, OptionVal }
...
+          } else {
+            // since instruments are sorted on both sides skip the serialized one and retry the local one
+            log.debug("Skipping serialized data in message for RemoteInstrument {} that has no local match", key)
+          }
+          buffer.position(nextPos)
+        }
+      } else {
+        log.debug(
+          "Skipping serialized data in message for RemoteInstrument(s) {} that has no local match",
+          remoteInstrumentIdIteratorRaw(buffer, endPos).mkString("[", ", ", "]"))
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala]: is this costly? perhaps add `if (log.isDebugEnabled)`?

**Review by patriknw — APPROVED** [review]
LGTM, minor nitpick

File: akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala (hunk 1 above)
Comment (bantonsson) [inline akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala]: Absolutely. I'll add some docs.

File: akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala (hunk 2 above)
Comment (bantonsson) [inline akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala]: You're right. The `mkString` is evaluated directly here. I'll fix it. I wonder why I was under the impression that it was call by name?

**Comment by patriknw** [comment]
Thanks, @bantonsson !

**Comment by patriknw** [comment]
Refs https://github.com/akka/akka/issues/21482

**Pull request merged**